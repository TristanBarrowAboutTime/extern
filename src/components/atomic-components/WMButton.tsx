import React, { FunctionComponent } from 'react';
import { createUseStyles } from 'react-jss';

/**
 * For now making sure that diabled state matches with the disabled styles is 
 * left to the parent. It is probably best to change that in the future.
 */


export enum ButtonType {
    NORMAL = 'NORMAL',
    GREEN = 'GREEN',
    RED = 'RED',
    NAKED = 'NAKED',
}

type WMButtonProps = {
    buttonType?: ButtonType,
    text: string,
    onClick: () => void,
    disabled?: boolean,
}

const btnBase = {
    padding: {
        top: 9, bottom: 8, left: 11, right: 12,
    },
    border: 0,
    '&:focus': {
        outline: 'none'
    }
}

const useStyles = createUseStyles({
    greenButton: {
        ...btnBase,
        backgroundColor: '#85B554',
        color: 'white',
    },
    redButton: {
        ...btnBase,
        backgroundColor: '#9B3E38',
        color: 'white',
    },
    normalButton: {
        ...btnBase,
        backgroundColor: '#333333',
        color: 'white',
    },
    nakedButton: {
        ...btnBase,
        backgroundColor: 'transparent',
        color: 'black',
    },
    disabledButton: {
        ...btnBase,
        backgroundColor: '#E9E9E9',
        color: '#999999',
    },
    nakedDisabledButton: {
        ...btnBase,
        backgroundColor: 'transparent',
        color: '#999999',
    }

});

export const WMButton: FunctionComponent<WMButtonProps> = ({
    buttonType = ButtonType.NORMAL, 
    text,
    onClick,
    disabled = false
}: WMButtonProps) => {
    const classes = useStyles();
    let buttonClass: string;
    
    switch (buttonType) {
        case ButtonType.GREEN:
            buttonClass = classes.greenButton;
            break;
        case ButtonType.RED: 
            buttonClass = classes.redButton;
            break;
        case ButtonType.NORMAL: 
            buttonClass = classes.normalButton;
            break;
        case ButtonType.NAKED: 
            buttonClass = classes.nakedButton;
            break;
        default:
            buttonClass = classes.normalButton;
    }
    if (disabled) {
        buttonClass = classes.disabledButton;
        if (buttonType === ButtonType.NAKED) buttonClass = classes.nakedDisabledButton;
    }

    return (
        <button 
            className={buttonClass}
            onClick={onClick}
            disabled={disabled}
        >
            {text}
        </button>
    );
}

export default WMButton;