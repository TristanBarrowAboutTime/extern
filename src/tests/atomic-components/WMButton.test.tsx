import React from 'react';
import { render } from '@testing-library/react';
import WMButton, { ButtonType } from '../../components/atomic-components/WMButton';

/*
 * for now snapshot testing should be sufficiant. 
 * it would be good to include click testing and className testing.
*/
describe('Button Displays Correctly', () => {

    /**
     * Base
     */
    test('that button displays with minimal props', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });

    /**
     * Normal
     */
    test('Normal buttons match snapshots', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NORMAL}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
    test('Normal buttons display disabled', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NORMAL}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });


    /**
     * Green
     */
    test('Green buttons match snapshots', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.GREEN}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
    test('Green buttons display disabled', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.GREEN}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });

    /**
     * Red
     */
    test('Red buttons match snapshots', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.RED}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
    test('Red buttons display disabled', () => {
        const {container} = render(
            <WMButton
                text='Button'
                buttonType={ButtonType.RED}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
    
    /**
     * Naked
     */
    test('Naked buttons match snapshots', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NAKED}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
    test('Naked buttons display disabled', () => {
        const {container} = render(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NAKED}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(container).toMatchSnapshot();
    });
});
