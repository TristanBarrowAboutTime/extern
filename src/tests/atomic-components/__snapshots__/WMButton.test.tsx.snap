// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Displays Correctly Green buttons display disabled 1`] = `
<div>
  <button
    class="disabledButton-0-2-5"
    disabled=""
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Green buttons match snapshots 1`] = `
<div>
  <button
    class="greenButton-0-2-1"
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Naked buttons display disabled 1`] = `
<div>
  <button
    class="nakedDisabledButton-0-2-6"
    disabled=""
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Naked buttons match snapshots 1`] = `
<div>
  <button
    class="nakedButton-0-2-4"
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Normal buttons display disabled 1`] = `
<div>
  <button
    class="disabledButton-0-2-5"
    disabled=""
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Normal buttons match snapshots 1`] = `
<div>
  <button
    class="normalButton-0-2-3"
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Red buttons display disabled 1`] = `
<div>
  <button
    class="disabledButton-0-2-5"
    disabled=""
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly Red buttons match snapshots 1`] = `
<div>
  <button
    class="redButton-0-2-2"
  >
    Button
  </button>
</div>
`;

exports[`Button Displays Correctly that button displays with minimal props 1`] = `
<div>
  <button
    class="normalButton-0-2-3"
  >
    Button
  </button>
</div>
`;
